core/driver/sources/src/main/scala/org/virtuslab/ideprobe/config/IntelliJProvider.scala

```diff
@@ -0,0 +1,93 @@
+package org.virtuslab.ideprobe.config
+
+import com.google.gson.Gson
+
+import java.nio.file.{CopyOption, FileVisitResult, Files, Path, SimpleFileVisitor}
+import org.virtuslab.ideprobe.ide.intellij.{InstalledIntelliJ, IntelliJFactory, LocalIntelliJ}
+import org.virtuslab.ideprobe.IdeProbePaths
+import org.virtuslab.ideprobe.dependencies.{IntelliJVersion, Plugin}
+
+import java.nio.file.attribute.BasicFileAttributes
+import scala.io.Source
+
+trait IntelliJProvider {
+  def factory: IntelliJFactory
+  def setup(): InstalledIntelliJ
+  def version: IntelliJVersion
+  def withFactory(factory: IntelliJFactory): IntelliJProvider
```

Comment:
method implementation is same in both child classes, can be implemented in here?

Reply:
This is tricky, as `withFactory` relies on `copy` method in case classes, therefore it's impossible to provide a completely subtype-agnostic implementation here. One possible idea would be something along this line:
```
final def withFactory(factory: IntelliJFactory): IntelliJProvider = this match {
  case any: ExistingIntelliJ => any.copy(factory = factory)
  case any: DefaultIntelliJ => any.copy(factory = factory)
}
```
however this solution's elegance is disputable.